### ISSUE TITLE
Timeout for one request makes the others time out too

### ISSUE THREAD

**Issue by lassebunk** [typhoeus/typhoeus#417]
First of all, thanks for creating Typhoeus. We (Bogrobotten) are using it to fetch hundreds of thousands of URLs every day. It really is a great library.

However, it seems like when we use Typhoeus::Hydra to fetch e.g. 50 URLs and one of them is taking a long time to respond, and we have a timeout of 10 seconds on each request, all of them time out, even though only one request was taking a long time to respond.

Is this the expected behavior, and could we do something to prevent a single slow request from affecting the other requests in the hydra?

Thanks in advance.

**Comment by PascalTurbo** [comment]
I got the same Problem - posted it on google-groups (bevor I've see this issue).

**Comment by i0rek** [comment]
Thanks for providing more informations @zt2 @PascalTurbo.
@PascalTurbo I think your understanding of timeouts is correct and I will look into that problem.

**Comment by revett** [comment]
We are seeing this problem as well. Tested with: 10 queued requests in `Hydra` to **S3**

**Comment by revett** [comment]
+1 please

We are seeing this problem as well. Tested with a loop: 

- 10 queued requests in `Hydra` to **S3**.
- Timeout of 2 seconds.
- Each request on average takes 150-250ms.

We see an exception thrown afte the test has been running for a short time. The strange thing is that it nearly always (95%) responds with a `status_code = 0` response, as seen in this issue: https://github.com/typhoeus/typhoeus/issues/411.

We're running **jruby-1.7.19**.

**Comment by Zapotek** [comment]
This is really messing me up too, if one request times out the whole queue does to hell.

**Comment by jpmoral** [comment]
Does this still happen?  I'd be willing to take a look if someone writes a spec for it.

**Comment by revett** [comment]
@jpmoral do you want a test `file.rb` that recreates the error?

**Comment by jpmoral** [comment]
@revett 
Sure, that would be helpful.

**Comment by jpmoral** [comment]
@revett

Here's what I used:
https://gist.github.com/jpmoral/1765e88f6dc526397660

I was able to reproduce the issue last week, I can't seem to now.  I used @zt2's code and changed it a bit.  I use the same method to test: kill the server for a while and then bring it back up.  A spec would be much better, I'm still working on one.

**Comment by Zapotek** [comment]
@jpmoral It'd be easier just putting a `#sleep` call in the Sinatra handler.

**Comment by jpmoral** [comment]
@Zapotek 
Yes, it would.  :)

**Comment by i0rek** [comment]
you can pass a `delay` for each request to the test server: https://github.com/typhoeus/typhoeus/blob/master/spec/support/server.rb#L81.

**Comment by nathanhoel** [comment]
Any progress on this? I was going to take a crack at it but I was happy to see work already underway.

**Comment by jpmoral** [comment]
@nathanhoel 
I tried to write a spec but I haven't been able to reproduce the issue again.  Do you have specs or code that reliably demonstrates it?

**Comment by nathanhoel** [comment]
No, I have not tried yet that is why I was asking about it to see if I should invest anytime in it.
I can see how this would be difficult to reproduce.

**Comment by Zapotek** [comment]
I'm getting it when running against remote servers but can't seem to reproduce it against a local sinatra test server. I'll keep trying.

**Comment by att14** [comment]
This is pretty serious. Here is an easy way to reproduce: https://gist.github.com/att14/edfceea56e434642aa1f

Whats the priority on this?

**Comment by i0rek** [comment]
@att14 thanks for reporting. This is the most serious bug I am aware of right now.

**Comment by att14** [comment]
Any news on this? I looked at the code and it seems to all be in Ethon. Is there a related ticket? If someone could describe the problem to me I could take a stab at fixing it.

**Comment by i0rek** [comment]
@att14 you are right, the problem is in Ethon. There is no related ticket yet. I didn't run into that myself. What do you want to know exactly?

**Comment by att14** [comment]
Well, I don't know anything about Ethon. Could you point me at what you believe to be the offending code?

**Comment by i0rek** [comment]
@att14 the hydra libcurl interaction mostly happens in https://github.com/typhoeus/ethon/blob/master/lib/ethon/multi/operations.rb.

**Comment by ptrippett** [comment]
I have done some testing on this issue today and think I have made some progress. 

The first thing I did was write a test based on code provided by @att14 which shows the error occurring by creating 200 connections to the local web server one being a timeout. Surprise, surprise it started failing

```
$ rspec spec/ethon/multi/operations_spec.rb:189
Run options: include {:locations=>{"./spec/ethon/multi/operations_spec.rb"=>[189]}}

Ethon::Multi::Operations
  #perform
    recovers from timeout
      only the request that timed out returns an error (FAILED - 1)

Failures:

  1) Ethon::Multi::Operations#perform recovers from timeout only the request that timed out returns an error
     Failure/Error: expect(easies.map(&:response_code).select(&:zero?).count).to eql 1
       
       expected: 1
            got: 155
       
       (compared using eql?)
     # ./spec/ethon/multi/operations_spec.rb:205:in `block (4 levels) in <top (required)>'

Finished in 0.17516 seconds
1 example, 1 failure

Failed examples:

rspec ./spec/ethon/multi/operations_spec.rb:204 # Ethon::Multi::Operations#perform recovers from timeout only the request that timed out returns an error
```

So 155 URLs failed to be fetched

The next thing I tried was reducing the number of urls to see if the test passed. When run with a pool of 50 urls and the result.

```
$ rspec spec/ethon/multi/operations_spec.rb:189
Run options: include {:locations=>{"./spec/ethon/multi/operations_spec.rb"=>[189]}}

Ethon::Multi::Operations
  #perform
    recovers from timeout
      only the request that timed out returns an error

Finished in 5.17 seconds
1 example, 0 failures
```

So why does this work with 50 urls but not 200. The answer is in the inner workings of libcurl multi, it seems that curl multi does no internal throttling of connections if you give it 1000 urls it will go off an simultaneously try to fetch all those urls. 

For each url in the queue we need a file descriptor available so lets see how many file descriptors i can use

```
$ ulimit -n
256
```

so what happens if we increase the value and rerun the test with 200 urls again

```
$ ulimit -n
256
$ ulimit -n 1024
$ ulimit -n
1024
$ rspec spec/ethon/multi/operations_spec.rb:189
Run options: include {:locations=>{"./spec/ethon/multi/operations_spec.rb"=>[189]}}

Ethon::Multi::Operations
  #perform
    recovers from timeout
      only the request that timed out returns an error

Finished in 1 minute 6.39 seconds
1 example, 0 failures
```

Voila!. So increasing the max file descriptors made the test pass. It appears as-is you need enough available file descriptors for the number of urls in the curl multi queue.

So whats the long term fix?

It appears [from this thread](http://curl.haxx.se/mail/lib-2010-06/0018.html) on the curl mailing list we will need to implement our own connection throttling inside Typheous so this is my next job!

Can anyone else confirm these findings?

**Comment by att14** [comment]
@ptrippett Awesome! I'm not able to fully confirm your findings, but it sounds reasonable. I did some testing a few days ago and from what I could tell the issue was inside of libcurl, but I was not brave enough to find out where.

**Comment by Zapotek** [comment]
@ptrippett: @att14's code triggers the issue for me, but I've got a `ulimit -n` of `1024`.

**Comment by i0rek** [comment]
@ptrippett thanks for you work!!

> The answer is in the inner workings of libcurl multi, it seems that curl multi does no internal throttling of connections if you give it 1000 urls it will go off an simultaneously try to fetch all those urls.

Thats not exactly what happens I belive. According to http://curl.haxx.se/libcurl/c/CURLMOPT_MAXCONNECTS.html it only creates a connection for every 4th easy: 

> By default libcurl will enlarge the size for each added easy handle to make it fit 4 times the number of added easy handles.

It is possible that I am interpreting this wrong. And maybe the details are irrelevant if you can help it by increasing the limits. Maybe there is more than one FD for every easy.

> Edit: Around this point I confused myself being in Ethon talking about queues forgetting Hydra is actually in Typheous!

Ya, I can see how that happened. The Hydra equivalent in Ethon is Multi.

**Comment by Zapotek** [comment]
Could it be that the [select](https://github.com/typhoeus/ethon/blob/master/lib/ethon/multi/operations.rb#L118) stuff need to track and adjust individual timeouts for connections (kind of like [this](https://github.com/rapid7/raptor-io/blob/master/lib/raptor-io/protocol/http/client.rb#L218)) instead of what's going on now?

**Comment by i0rek** [comment]
@Zapotek the example you found seems to be a timeout for checking a socket. In libcurl though the timeouts are more - like the connect_timeout. It is impossible for Ethon to do that manually.

**Comment by ptrippett** [comment]
@Zapotek Your right it does still fail. I had to modify the server slightly because of io_fillbuf errors in webrick which I dont think are part of this same problem so I replaced webrick with thin

```ruby
require "sinatra"
require "thin"

set :server, 'thin'

get "/" do
  "Hello world!"
end

get "/timeout" do
  loop {}
end
```

I then started paying with the `max_connections` option to hydra.

```ruby
require_relative "./lib/typhoeus" # I am requiring relative because I am debugging

# Setting (max_concurrency: 20) here depending on the FD's available
#
hydra = Typhoeus::Hydra.new(max_concurrency: 20)

requests = 1000.times.map { Typhoeus::Request.new('localhost:4567', method: :get, timeout: 5) }
requests << Typhoeus::Request.new('localhost:4567/timeout', method: :get, timeout: 5)

requests.each { |request| hydra.queue request }

hydra.run

requests.each do |request|
  if request.response.code != 200
    p "ERROR #{request.response.code}"
  end
end
```

I started incrementing the `max_connections` variable on each run and I seem to be ok up to ~100 max connections after that I start seeing odd logs coming from thin.

Successful request:

```
127.0.0.1 - - [27/Sep/2015:13:35:30 +0200] "GET / HTTP/1.1" 200 12 2.3658
```

Unsuccessful request:

```
- - - [27/Sep/2015:13:35:31 +0200] "GET / HTTP/1.1" 200 12 0.7198
```

Notice theres no requesting host in the failed log line but but the servers returning a 200 status.

**Comment by Zapotek** [comment]
I was using `thin` in my test and failures do increase with the `max_concurrency`.
I was getting errors with a concurrency of `20`, both in the test and in real world applications -- `10` seemed to do fine.

**Comment by Zapotek** [comment]
About the failing log line, the connection must have been closed after the request was received by the server, way before the timeout was reached though, since the response took only `0.7198`.
I'm just guessing.

**Comment by wflanagan** [comment]
Is there any update on this?

**Comment by danielmiessler** [comment]
So this basically leaves us with massive numbers of false responses, and there's been no update on this in over two years?

**Comment by wflanagan** [comment]
Nope.  Not that i’m aware of.  Still somewhat busted.

**Comment by i0rek** [comment]
A new ethon was released with a bugfix for a similar problem, maybe it fixes this one as well: https://github.com/typhoeus/ethon/pull/125.

**Comment by dblock** [comment]
I believe this is resolved with recent versions of ethon.

**Comment by felipedmesquita** [comment]
Thanks @lassebunk for your report and everyone who commented here, especially @721p for the reproducible script and server. I've run it with the latest version and only the requests that are supposed to timeout do so. This seems to be resolved, I'm closing it and ask that any of you open a new issue if you still run into problems for a recent version of Typhoeus.
